### ISSUE TITLE
[Rush] The NPM shrinkwrap file does not provide gulp.

### ISSUE THREAD

**Issue by MartynasZilinskas** [Microsoft/web-build-tools#257]
_Output:_
```sh
$ rush install

Rush Multi-Package Build Tool 3.0.9 - http://aka.ms/rush


Starting "rush install"

Found NPM version 5.0.3 in C:\Users\QuatroDev\.rush\npm-5.0.3

Symlinking "D:\SimplrJS\simplr-forms\common\temp\npm-local"
  --> "C:\Users\QuatroDev\.rush\npm-5.0.3"

Updating temp projects in D:\SimplrJS\simplr-forms\common\temp\projects

The NPM shrinkwrap file does not provide "gulp" (github:gulpjs/gulp#4.0) required by pinned versions
Updating D:\SimplrJS\simplr-forms\common\temp\npm-shrinkwrap.json

The NPM shrinkwrap file is missing "gulp" (github:gulpjs/gulp#4.0) required by "@simplr/react-forms".

The NPM shrinkwrap file is missing "gulp" (github:gulpjs/gulp#4.0) required by "@simplr/react-forms-dom".

The NPM shrinkwrap file is missing "gulp" (github:gulpjs/gulp#4.0) required by "@simplr/react-forms-validation".

You need to run "rush generate" to update your NPM shrinkwrap file.
```

I ran into problem when using dependency from github at specific tag in my packages.
```
    "gulp": "github:gulpjs/gulp#4.0",
```

From `npm-shrinkwrap.json`:
```
    "gulp": {
      "version": "github:gulpjs/gulp#38246c3f8b6dbb8d4ef657183e92d90c8299e22f",
```
It takes last commit hash from `4.0` branch.

**Comment by pgonzal** [comment]
Does this problem persist even after you run "`rush generate`"?  If so, could you provide some simple repro steps?  Thanks!

**Comment by pgonzal** [comment]
I see.  The problem in this case is that [https://github.com/Microsoft/web-build-tools/blob/5513858b1d478bd5670e9a738936b0e87a88aab3/apps/rush/src/utilities/ShrinkwrapFile.ts#L88](ShrinkwrapFile.hasCompatibleDependency()) only considers SemVer compatibility:

```JavaScript
    // If we found it, the version must be compatible
    return semver.satisfies(dependencyJson.version, versionRange);
```

But in your case, the version pattern is  `"github:gulpjs/gulp#4.0"` not a SemVer expression.  If we support this syntax, we would need to determine that `"#4.0"` and `"#38246c3f8b6dbb8d4ef657183e92d90c8299e22f"` are equivalent somehow.  Would that require Rush to make a network request to GitHub?  I wonder how NPM handles it.

It's probably not too difficult to implement, but there are a number of places where Rush performs these sorts of comparisons, so we would need to test it thoroughly.  Since this is a feature request, we should prioritize it based on demand, versus the difficulty of implementing it.

How important is this scenario to you?  There are a number of other features that currently have pretty high demand, e.g. cross-repo linking, better incremental builds, [pnpm](https://www.npmjs.com/package/pnpm) support, etc

**Comment by MartynasZilinskas** [comment]
Hey @pgonzal.

We are currently using rush in one of our open-source projects [Link to repo](https://github.com/SimplrJS/react-forms).
With this issue, we must use `rush generate` instead of `rush install` to install packages. I think it is rush's bug, since it works with `npm install`, even when `npm-srinkwrap.json` is present.

Me and @DovydasNavickas were searching npm source code to see how npm resolves these kind of dependencies. Here's what we found: https://github.com/npm/npm/blob/90a25f65804d7d7dc351d01c53826b879139d8da/lib/install/deps.js#L39. 
I think that will be helpful.

**Comment by pgonzal** [comment]
The Rush operation that I described will examine your [react-forms/package.json](https://github.com/SimplrJS/react-forms/blob/master/packages/react-forms/package.json)...

```
  ...
  "enzyme": "^2.8.2",
  "gulp": "github:gulpjs/gulp#4.0",
  "jest": "^20.0.4",
  ...
```

...and then compare it to your [npm-shrinkwrap.json](https://github.com/SimplrJS/react-forms/blob/master/common/config/rush/npm-shrinkwrap.json)...

```
  ...
  "gulp": {
    "version": "github:gulpjs/gulp#38246c3f8b6dbb8d4ef657183e92d90c8299e22f",
    "dependencies": {
  ...
```
...in order to determine whether the "gulp" version described by the shrinkwrap will satisfy the request in your package.json file.  If not, Rush will ask you to run "`rush generate`" (presumably because somebody has modified package.json after the last "`rush generate`").

I was saying that in order to answer this question accurately, we would need to make a REST API call to github.com to determine whether the shrinkwrapped MD5 hash refers to the "4.0" Git branch or not.  I didn't look very closely, but if I understand right, the NPM source code that you referenced takes a lazier approach:  If the version spec is not SemVer syntax, then NPM simply assumes that it's incompatible, and forces a reinstall every time:

```javascript
  try {
    return semver.satisfies(child.package.version, requested.fetchSpec)
  } catch (e) {
    return false
  }
```

If we did this, we would force you to run "`rush generate`" every time, as you are already doing.

Here's another possibility:  What if Rush made no attempt to validate "`github:`" references?  Instead, "`rush install`" would happily accept them, and it's up to you to figure out whether you need to run "`rush generate`".

The bad user experience would be like this:
1. Someone on your team modifies package.json to request to "`github:gulpjs/gulp#4.1`" instead of "`4.0`"
2. They run "`rush install`" -- everything seems to be cool
3. They complain that for some strange reason, Rush is still installing the 4.0 branch.
4. You explain that they need to run "`rush generate`" for this sort of change (but in general, they should NOT run "`rush generate`" every time they modify package.json)

This isn't great, but it's not too bad, and pretty easy to implement.

The ideal would be for "`rush generate`" to somehow store the mapping from "github:gulpjs/gulp#4.0" --> "github:gulpjs/gulp#38246c3f8b6dbb8d4ef657183e92d90c8299e22f", to solve the root problem that `npm-shrinkwrap.json` is missing this information.  That would avoid the bad user experience, but at a price of more complexity and a more complicated implementation.

**Comment by pgonzal** [comment]
FYI I just tried an experiment with NPM:

1. Create a package.json with "`github:gulpjs/gulp#master`"
2. Run "`npm install`"
3. Delete the package-lock.json file
4. Modify package.json to say "`github:gulpjs/gulp#4.0`"
5. Run "`npm install`" again

It seems that `npm install` does NOT force a reinstall of gulp.  Instead, it happily says that everything is okay, leaving you with the wrong gulp version in your node_modules folder.

So if we implement this hack for Rush, it will be no worse than the behavior of NPM itself.  (Not that NPM is any reference point for great software engineering heheh.)  :-)

**Comment by pgonzal** [comment]
@MartynasZilinskas Could you test this PR and let us know if it solves your problem?

**Comment by MartynasZilinskas** [comment]
Yeah, it solves the problem. It emits a warning as intended. :+1:

```sh
Rush Multi-Package Build Tool 3.0.9 - http://aka.ms/rush



Starting "rush install"

Found NPM version 5.0.3 in C:\Users\QuatroDev\.rush\npm-5.0.3

Symlinking "D:\SimplrJS\simplr-forms\common\temp\npm-local"
  --> "C:\Users\QuatroDev\.rush\npm-5.0.3"

Updating temp projects in D:\SimplrJS\simplr-forms\common\temp\projects
WARNING: Not validating git-based spec: "github:gulpjs/gulp#4.0"
Updating D:\SimplrJS\simplr-forms\common\temp\npm-shrinkwrap.json

Checking node_modules in D:\SimplrJS\simplr-forms\common\temp

Deleting the NPM cache folder
Deleting the "npm-tmp" folder

Running "npm install" in D:\SimplrJS\simplr-forms\common\temp


> fsevents@1.1.2 install D:\SimplrJS\simplr-forms\common\temp\node_modules\fsevents
> node install


> uglifyjs-webpack-plugin@0.4.6 postinstall D:\SimplrJS\simplr-forms\common\temp\node_modules\uglifyjs-webpack-plugin
> node lib/post_install.js

added 907 packages in 65.734s

Done. (1 minute 6.6 seconds)
Starting "rush link"

LINKING: @simplr/mvdir
Purging D:\SimplrJS\simplr-forms\tools\mvdir\node_modules

LINKING: @simplr/react-forms
Purging D:\SimplrJS\simplr-forms\packages\react-forms\node_modules

LINKING: @simplr/react-forms-dom
Purging D:\SimplrJS\simplr-forms\packages\react-forms-dom\node_modules

LINKING: @simplr/react-forms-validation
Purging D:\SimplrJS\simplr-forms\packages\react-forms-validation\node_modules
Writing "D:\SimplrJS\simplr-forms\common\temp\rush-link.json"

Rush link finished successfully. (4.13 seconds)

Next you should probably run: "rush rebuild"
```

**Comment by pgonzal** [comment]
Fixed in `@microsoft/rush@3.0.11`
